url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/screen/Screen.java

```diff
@@ -0,0 +1,490 @@
+    	for(int i = 0; i < string.length(); i++)
+    		putCharacter(x + i, y, 
+                        new ScreenCharacter(string.charAt(i), 
+                                            foregroundColor, 
+                                            backgroundColor,
+                                            styles));
+    }
+
+    void putCharacter(int x, int y, ScreenCharacter character)
+    {
+        synchronized(mutex) {
+            if(y < 0 || y >= backbuffer.length || x < 0 || x >= backbuffer[0].length)
+                return;
+
+            //Only create a new character if the 
+            if(!backbuffer[y][x].equals(character))
+                backbuffer[y][x] = new ScreenCharacter(character);
+        }
+    }
+
+    /**
+     * This method will check if there are any resize commands pending. If true, 
+     * you need to call refresh() to perform the screen resize
+     * @return true if the size is the same as before, false if the screen should be resized
+     */
+    public boolean resizePending()
+    {
+        synchronized(resizeQueue) {
+            return !resizeQueue.isEmpty();
+        }
+    }
+
+    /**
+     * Call this method to make changes done through {@code putCharacter(...)},
+     * {@code putString(...)} visible on the terminal. The screen will calculate
+     * the changes that are required and send the necessary characters and
+     * control sequences to make it so.
+     */
+    public void refresh()
+    {
+        if(!hasBeenActivated)
+            return;
+        
+        synchronized(mutex) {
+            //If any resize operations are in the queue, execute them
+            resizeScreenIfNeeded();
+
+            Map<TerminalPosition, ScreenCharacter> updateMap = new TreeMap<TerminalPosition, ScreenCharacter>(new ScreenPointComparator());
+            
+            for(int y = 0; y < terminalSize.getRows(); y++)
+            {
+                for(int x = 0; x < terminalSize.getColumns(); x++)
+                {
+                    ScreenCharacter c = backbuffer[y][x];
+                    if(!c.equals(visibleScreen[y][x]) || wholeScreenInvalid) {
+                        visibleScreen[y][x] = c;    //Remember, ScreenCharacter is immutable, we don't need to worry about it being modified
+                        updateMap.put(new TerminalPosition(x, y), c);
+                    }
+                }
+            }
+
+            Writer terminalWriter = new Writer();
+            terminalWriter.reset();
+            TerminalPosition previousPoint = null;
+            for(TerminalPosition nextUpdate: updateMap.keySet()) {
+                if(previousPoint == null || previousPoint.getRow() != nextUpdate.getRow() ||
+                        previousPoint.getColumn() + 1 != nextUpdate.getColumn()) {
+                    terminalWriter.setCursorPosition(nextUpdate.getColumn(), nextUpdate.getRow());
+                }
+                terminalWriter.writeCharacter(updateMap.get(nextUpdate));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.